Fixing the Item height for Table and Tree when the default font is set

elsazac opened the pull request:
Fixes: #677  

A new function named as `getInset()` is created in `Table.java`. This function calculates the inset value of a cell in a table for aarch64 architecture. The inset value of a row in a table is calculated by subtracting the height of a cell from a frame.This calculated inset value is added to the height in `setItemHeight()` function.

The inset value of 1 is retained for x86_64 architecture.

/cc @lshanmug

Phillipus commented:
Hi, I tested the patch. If there are no items the heights are too small:

<img width="612" alt="s" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/600504/66611bb4-1fa7-4633-a8a2-8dc7e64ada1b">

```java
import org.eclipse.jface.layout.GridDataFactory;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.swt.widgets.Table;
import org.eclipse.swt.widgets.Tree;

public class FontTest {

    public static void main(String[] args) {
        Display display = new Display();
        Shell shell = new Shell(display);
        shell.setText("Tree & Table Font");
        shell.setLayout(new GridLayout());
        
        Tree tree = new Tree(shell, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
        GridDataFactory.create(GridData.FILL_BOTH).applyTo(tree);
        tree.setLinesVisible(true);
        
        Table table = new Table(shell, SWT.BORDER | SWT.V_SCROLL | SWT.H_SCROLL);
        GridDataFactory.create(GridData.FILL_BOTH).applyTo(table);
        table.setLinesVisible(true);
        
        shell.setSize(500, 450);
        shell.open();
        
        tree.setFont(tree.getFont());
        table.setFont(table.getFont());
        
        while(!shell.isDisposed()) {
            if(!display.readAndDispatch()) display.sleep();
        }
        
        display.dispose();
    }
}
```

elsazac commented:
@Phillipus 

Can you test this patch?

Phillipus commented:
> @Phillipus
> 
> Can you test this patch?

Hi @elsazac,

I've tested the patch on Mac M1 Ventura 13.6 and it does fix the row height problem testing with the Snippet in #771 and also in our RCP application.

I remain puzzled, though, how and where the initial row height is set (25 pixels) before setting the font of the table or tree.

elsazac commented:
> @elsazac The problem also occurs for the `List` component. I've added a Snippet to the original issue [here](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1746844294). This will require a similar fix at line 1182 here:
> 
> https://github.com/eclipse-platform/eclipse.platform.swt/blob/17dcab01f437e21052189dc5cd394856868ad197/bundles/org.eclipse.swt/Eclipse%20SWT/cocoa/org/eclipse/swt/widgets/List.java#L1178-L1184

@Phillipus I have made a similar change in List as well.
Could you test it?

Phillipus commented:
> @Phillipus I have made a similar change in List as well. Could you test it?

Hi @elsazac - I tested and the fix works OK, thank-you.

But before committing this patch I have a doubt concerning what should be the true initial row height. See comment here - https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1750292461

elsazac commented:
> > @Phillipus I have made a similar change in List as well. Could you test it?
> 
> Hi @elsazac - I tested and the fix works OK, thank-you.
> 
> But before committing this patch I have a doubt concerning what should be the true initial row height. See comment here - [#677 (comment)](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1750292461)

@Phillipus I have responded in the linked issue. Can you review and help progress this PR?

merks commented:
It looks like this current incarnation sprinkles the appropriate magic numbers to fix the problem.  I think https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1846595468 answers the outstanding problem.

@Phillipus Are there any outstanding concerns with this PR?

Phillipus commented:
> @Phillipus Are there any outstanding concerns with this PR?

@merks Yes.

The larger item height is seen on macOS `aarch64` but not on `x86_64` when Eclipse or an RCP app is run from its launcher. As explained by @sravanlakkimsetti in https://bugs.eclipse.org/bugs/show_bug.cgi?id=574932#c2:

>  This is because eclipse launcher SDK version
>  On x86_64 SDK used is 10.14
>  On aarch64 SDK used is 11
>  But the way forward is to use latest SDK and update SWT to MacOS 11 SDK.

However, I recently installed a new JDK on an Intel Mac (x86_64) that has been compiled against Mac SDK 11.1 and when launching a child Eclipse, or our RCP app, the item height is also bigger. This is because when running a child Eclipse it uses the `java` binary, not the launcher, and this one has been compiled using 11.1 SDK.

So the inconsistency is due not to OS architecture but on the linked SDK version of the launcher or the `java` binary and perhaps we should consider this needs to be addressed by rebuilding the `x86_64` launcher against MacOS 11 SDK:

>  But the way forward is to use latest SDK and update SWT to MacOS 11 SDK.

I don't know how that is done (is it the equinox binaries project?)

If this is done, however, note that Eclipse and RCP apps will not be able to run on macOS < 11 (before Big Sur)

merks commented:
I see!  Thank you for the detailed information!!  Magic numbers never do feel quite right.  😱

sravanlakkimsetti commented:
> I don't know how that is done (is it the equinox binaries project?)

I prefer not to have different code for x86 and arm implementation. My suggestion would be to use same masOS sdk for both versions of launcher. 

[Here is the section from equinox project](https://github.com/eclipse-equinox/equinox/blob/master/features/org.eclipse.equinox.executable.feature/library/cocoa/build.sh#L66-L74) where we switch between macos 11 and macos 10.14 sdk for arm and intel architectures

Removing this if condition and and rebuilding launcher should make make both versions of launcher work in the same way.

If you need to upgrade to latest version of macos sdk you'll need to work with foundation to get it upgraded on the [macos build machine](https://ci.eclipse.org/releng/computer/nc1ht-macos11-arm64)

Phillipus commented:
@sravanlakkimsetti Thanks for that information!

I notice that you did already do this here:

https://github.com/eclipse-equinox/equinox/commit/0dad4635cce80cdc9a16764b8470b56ee421d058

but reverted this:

https://github.com/eclipse-equinox/equinox/commit/4db53c06417fc1c72955a1477c4f9cf658e45f91

Did you revert because it was no longer needed die to https://github.com/eclipse-platform/eclipse.platform.swt/issues/452 fixing itself?

Phillipus commented:
Opened https://github.com/eclipse-equinox/equinox/issues/473

sravanlakkimsetti commented:
> @sravanlakkimsetti Thanks for that information!
> 
> I notice that you did already do this here:
> 
> [eclipse-equinox/equinox@0dad463](https://github.com/eclipse-equinox/equinox/commit/0dad4635cce80cdc9a16764b8470b56ee421d058)
> 
> but reverted this:
> 
> [eclipse-equinox/equinox@4db53c0](https://github.com/eclipse-equinox/equinox/commit/4db53c06417fc1c72955a1477c4f9cf658e45f91)
> 
> Did you revert because it was no longer needed due to #452 fixing itself?

Not exactly. The problem was slightly complicated. At that time the supported java version was built with 10.14. When we moved to Macos 11, Eclipse had one behavior and the child, eclipse launched using run menu, had different behavior. For this reason we wanted to wait till java moved to Macos 11 before we move. 

Now that java also moved to Macos 11 we can move mow.  Lets reinstate https://github.com/eclipse-equinox/equinox/commit/0dad4635cce80cdc9a16764b8470b56ee421d058

Phillipus commented:
> Lets reinstate https://github.com/eclipse-equinox/equinox/commit/0dad4635cce80cdc9a16764b8470b56ee421d058

Could you do that?

sravanlakkimsetti commented:
> > Lets reinstate [eclipse-equinox/equinox@0dad463](https://github.com/eclipse-equinox/equinox/commit/0dad4635cce80cdc9a16764b8470b56ee421d058)
> 
> Could you do that?

Sure will take it up

Phillipus commented:
> Now that java also moved to Macos 11

For the Temurin version of Java this was a very recent move.

Temurin Java 17 moved to SDK 11.1 with version [jdk-17.0.9+9](https://github.com/adoptium/temurin17-binaries/releases/tag/jdk-17.0.9%2B9) - October 19 2023

Temurin Java 21 moved to SDK 11.1 with version [jdk-21.0.1+12](https://github.com/adoptium/temurin21-binaries/releases/tag/jdk-21.0.1%2B12) - October 24 2023

Phillipus commented:
Something to be aware of regarding using SDK 11 on the macOS Intel launcher in the future:

This patch will need to change from:

```java
int height = (int)Math.ceil (ascent + descent) + ((OS.IS_X86_64) ? 1 : 8);
```
to

```java
int height = (int)Math.ceil (ascent + descent) + 8;
```

And this means that an end user will also see the increased item height if they are using an older JDK that was compiled with 10.14.

Phillipus commented:
It turns out that the Mac x86_64 launcher has been built with SDK 13 since around Eclipse 4.30. So we don't need to wait.

@elsazac Could you update your patch by removing the `OS.IS_X86_64` condition so it applies to both x86_64 and aarch64, please?

elsazac commented:
> @elsazac Could you update your patch by removing the `OS.IS_X86_64` condition so it applies to both x86_64 and aarch64, please?

done.

Phillipus commented:
> done.

Thanks. I tested this on both x86_64 and aarch64 and it works OK.

However, a user will see discrepancies in Item height if they are launching a child Eclipse/RCP app using an older Java version (one compiled with Mac SDK 10.14). I guess in this case they should update to latest version.

I'm happy with this PR now. I don't know who should commit it.

Phillipus commented:
Is this ready to be merged now?

Phillipus commented:
Be aware that because of the increased item height in Lists, Tables and Trees, text is not vertically cantered in the `List` component (run this [Snippet](https://github.com/eclipse-platform/eclipse.platform.swt/issues/677#issuecomment-1746844294) to see) nor in Cell Editors (`TableEditor` and `TreeEditor`). It looks kind of ugly but I guess that would need further investigation (I tried but failed)

Phillipus commented:
@elsazac In three places a value of 8 is added (it was 1). Do you think it would be a good idea to create a static constant in each class for this? Something like the existing `static final int CELL_GAP = 1` but with a new name? And add a comment to explain what it is for?

elsazac commented:
> @elsazac In three places a value of 8 is added (it was 1). Do you think it would be a good idea to create a static constant in each class for this? Something like the existing `static final int CELL_GAP = 1` but with a new name maybe `CELL_PADDING`)? And add a comment to explain what it is for?

I have added the changes and squashed the commits. Please have a look.

merks commented:
Nit picky detail... The comment `Row height for xxx item` is not really a correct/accurate description.  This is padding that is added to the height as determined by the font metrics, so `Row height padding for ...`

elsazac commented:
> Nit picky detail... The comment `Row height for xxx item` is not really a correct/accurate description. This is padding that is added to the height as determined by the font metrics, so `Row height padding for ...`

I have updated the comments.

merks commented:
Yes, that's a better more descriptive name too!  😄

Phillipus commented:
Thanks @elsazac for your work on this, and thanks @merks for the feedback and merge. :-)
